### PR TITLE
Delete GCR Images

### PR DESCRIPTION
**Pull request by colerogers**
### Description

An open-sesame command for purging the contents of the `gcf/` directory in every GCR subdomain. The `gcf/` directory is created by Google Cloud Functions on a function deployment. GCF places function images inside this directory, a user should not create custom images inside this directory. To limit misuse of this command, it requires the open-sesame preview flag and displays all artifacts that will be deleted during the confirmation step. The `force` flag is not enabled for the command.

### Scenarios Tested

- Invalid region
- Searching single region
- Searching multiple regions
- Searching multiple subdomains
- Delete single regions
- Delete multiple regions
- Delete multiple subdomains

### Sample Commands

`firebase --open-sesame deletegcfartifacts`

`firebase functions:deletegcfartifacts`

`firebase --close-sesame deletegcfartifacts`

### CODE REVIEW

File: src/deploy/functions/containerCleaner.ts
Comment (joehan) [inline src/deploy/functions/containerCleaner.ts:136]: Prefer describing the behavior in a little more detail , and documenting any non-obvious behavior (like what happens if location is omitted). Something like:
```
/**
* Deletes all artifacts from GCF folder in GCR.
* @param projectId
* @param location: the subdomain to clean up. If omitted, will delete all locations.
*/
```

File: src/deploy/functions/containerCleaner.ts
Comment (joehan) [inline src/deploy/functions/containerCleaner.ts:145]: I don't think we should make this a class - theres no constructor, and there's only one real method, so this seems like it would be more readable as a plain old function.

File: src/previews.ts
Comment (joehan) [inline src/previews.ts:22]: This will all be subject to API council changes, but IMO, the open-sesame flag should just match whatever the name of the command is in this case.

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (joehan) [inline src/test/deploy/functions/containerCleaner.spec.ts:250]: Prefer a little more specificity here - you could check that it was specifically called with the ocrrect subdomains, for example.

**Review by joehan — COMMENTED** [review]
Took a quick pass over the non user facing bits - please rerequest review from me once API is approved.

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: Added TSDoc comments to all functions

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: Changed it to an exportable function. Added a default parameter for DockerHelper for unit testing

File: src/previews.ts
Comment (colerogers) [reply src/previews.ts]: Change to deletegcfartifacts per API review

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (colerogers) [reply src/test/deploy/functions/containerCleaner.spec.ts]: Updated all unit tests

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [inline src/deploy/functions/containerCleaner.ts:216]: This is using a default value so we can inject a stub for unit testing.

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [inline src/deploy/functions/containerCleaner.ts:293]: Same here - This is using a default value so we can inject a stub for unit testing.

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (inlined) [inline src/test/deploy/functions/containerCleaner.spec.ts:540]: extra line?

**Review by inlined — CHANGES_REQUESTED** [review]
Lots of through testing and documentation! I think it's worth questioning whether the amount of detail we're giving customers before deletion is useful/too costly. I'd play with console.cloud.google.com/gcr to see what it looks like yourself. As mentioned in comments, "digests" are the actual images, "children" are folders for images, and "tags" are freeform references to those "digests" that pin them from being deleted.

File: src/commands/functions-deletegcfartifacts.ts
Comment (colerogers) [reply src/commands/functions-deletegcfartifacts.ts:8]: @inlined  I'm okay with listing out the full GCR path since that would allow the user to easily check for artifacts without looking up UUIDs. This requirement came from a discussion with @kaibolay on the API proposal. Kai ultimately approved the proposal, so we should probably get his opinion on this change.

Comment (kaibolay) [reply src/commands/functions-deletegcfartifacts.ts:8]: I really like listing the full GCR path(s).

Comment (inlined) [reply src/commands/functions-deletegcfartifacts.ts:8]: I guess it's time to disagree yet commit =D

Comment (kaibolay) [reply src/commands/functions-deletegcfartifacts.ts:8]: Sorry, I should have been clearer. I don't think we disagree. 

What I meant to say is that I really like displaying the full directory path (without the UUIDs of the individual images).

Maybe we could display the number of images in those directories; e.g.:

```
You are about to delete all images in the following directories:

us.gcr.io/myproject/gcf/us-central1 (3 images)
us.gcr.io/myproject/gcf/us-west1 (1 image)

Are you sure?
```

But that's only a nice to have. The directory paths alone contain 99.9% of the important information.

File: src/commands/functions-deletegcfartifacts.ts
Comment (colerogers) [reply src/commands/functions-deletegcfartifacts.ts]: Changed

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: Removed this function, refactored for easier readability

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: Removed this function, refactored for easier readability

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: Removed this function, refactored for easier readability

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: Changed to `!SUBDOMAIN_MAPPING[location]`

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: Fixed

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: Changed

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (colerogers) [reply src/test/deploy/functions/containerCleaner.spec.ts]: Fixed

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (colerogers) [reply src/test/deploy/functions/containerCleaner.spec.ts]: Fixed

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (colerogers) [reply src/test/deploy/functions/containerCleaner.spec.ts]: Yep, fixed.

File: src/commands/functions-deletegcfartifacts.ts
Comment (colerogers) [reply src/commands/functions-deletegcfartifacts.ts]: fixed

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: The linter seems to place longer lambdas on their own line. I went through and changed them to blocks.

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (colerogers) [reply src/test/deploy/functions/containerCleaner.spec.ts]: Removed

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (colerogers) [reply src/test/deploy/functions/containerCleaner.spec.ts]: Changed the tests and added in digests & tags.

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: Thanks for writing this all out! I've changed the list function to use  what you wrote here. I modified the parallel work to go over the subdomains since we only need to run a `ls` at the root of the `gcf` directory. I also changed to helper to accept a subdomain instead of a location.

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: After refactoring the code, I ended up removing all of these helper functions. I kept the TSDocs on the two main exportable functions.

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [inline src/deploy/functions/containerCleaner.ts:17]: I added an export here so I can use the actual object in unit testing for deleting every region. I didn't want to create a 2nd source of truth and have failures if this ever gets updated

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [inline src/deploy/functions/containerCleaner.ts:164]: Looping through the subdomains here and filtering later since that limits us to calling the docker API only 3 times.

File: src/commands/functions-deletegcfartifacts.ts
Comment (colerogers) [reply src/commands/functions-deletegcfartifacts.ts]: I've changed it to display the full directory path. I explored adding in the image counts but that is a recursive operation that calls the docker API for each directory and sub directories. We would probably get rate limited and could potently run into a similar issue as [this](https://github.com/firebase/firebase-tools/issues/3639) where the command fails.

Thomas also pointed out that some customers that are blocked on this issue have a massive amount of images. If those customers were to run this command then they'd likely not make much progress waiting for the list to load or get fully blocked when running this command.

File: src/commands/functions-deletegcfartifacts.ts
Comment (inlined) [inline src/commands/functions-deletegcfartifacts.ts:23]: on Google Container Registry

File: src/deploy/functions/containerCleaner.ts
Comment (inlined) [inline src/deploy/functions/containerCleaner.ts:217]: Am I crazy here or are we creating a set from an array and then only using that set to create another array below (for map)

File: src/deploy/functions/containerCleaner.ts
Comment (inlined) [inline src/deploy/functions/containerCleaner.ts:218]: arrays are idiomatic in JavaScript and you're only appending to the end. You can use an array and just use push(). Array insertion into an end is O(1) when using the doubling method.

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (inlined) [inline src/test/deploy/functions/containerCleaner.spec.ts:239]: Nit: checking for it throwing is fragile because it only passes if the error is thrown before any await command. expect it to reject instead.

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (inlined) [inline src/test/deploy/functions/containerCleaner.spec.ts:262]: FYI. You can use `deep` here (though you'd have to use `sort` if the list has multiple elements.

```
expect(paths).to.deep.equal(["us.gcr.io/projects/gcf/us-central1"]);
```

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (inlined) [inline src/test/deploy/functions/containerCleaner.spec.ts:346]:
```
await expect(() => containerCleaner.deleteGcfArtifacts("project", undefined, helpers)).to.reject
```

(make your method async)

**Review by inlined — APPROVED** [review]

File: src/commands/functions-deletegcfartifacts.ts
Comment (colerogers) [reply src/commands/functions-deletegcfartifacts.ts]: whoops, fixed

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts:223]: I was using a set to remove duplicate values and used an array to keep consistency with the other delete function. I just switched to the `for ... of ...` syntax and `push` each `Stat` into a `Promise` array.

File: src/deploy/functions/containerCleaner.ts
Comment (colerogers) [reply src/deploy/functions/containerCleaner.ts]: Changed to an array.

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (colerogers) [reply src/test/deploy/functions/containerCleaner.spec.ts]: Changed to use `deep` in all occurrences.

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (colerogers) [reply src/test/deploy/functions/containerCleaner.spec.ts]: Fixed

File: src/test/deploy/functions/containerCleaner.spec.ts
Comment (colerogers) [reply src/test/deploy/functions/containerCleaner.spec.ts]: Fixed

**Pull request merged**
